### PR TITLE
Encoder-Decoder (T5) support

### LINKED ISSUE #258: [Proposal] Support BERT Model Series (and/or T5).
Screenshot 2023-04-26 at 8 00 40 pm

### Proposal 

A glaring gap in models supported by TransformerLens is the lack of support for encoder/decoder models like BERT and T5. In order to rectify this, we need to build support for encoder-decoder models, cross attention and likely a few other details. 

### Motivation

- I've been asked a number of times by people whether we have T5 or BERT support coming. 
- I think this is a reasonable contribution for people skilling up. You can ask for help or collaborate with other people working on building it. 
- I (Joseph) and likely others will be able to give feedback on your work and pitch in if needed! 

### Pitch

In order to complete this card we are going to build a class called HookedEncoderDecoder which can be instantiated and have the BERT or T5 weights from [huggingface](https://huggingface.co/bert-base-uncased) loaded in. This is the same strategy the package uses for other models so there's nothing fundamentally new, only components we haven't yet built with hooks. Once we have a class that has corresponding weights as needed, we will want to ensure all components are "hooked" correctly, enabling methods like .run_with_cache to work. It's not as complicated as it sounds to get it working. 

Possibly the "best" version of this is tricky but we're willing to tolerate some code duplication as long we're confident the implementation is accurate (ie: tests show the hooked activations are correct and inference is identical to what we get without using TransformerLens. 

Tasks:

- [ ] Make a fork for BERT support (keep it synced if building this takes more than a moment)
- [ ] Write some tests to be used during development, these should cover any novel architecture components these will involve during development. I'd start with two placeholder tests. 1. for do you have a model architecture which you can load weights into. and 2. if you had that architecture, would it produce identical results to the hugging-face model. 
- [ ] Load the BERT model, look at each component. Maybe read the BERT paper if needed to get context. Make a list of every nn.module that we don't currently support. Some items that will be on the list. 1. a linear block before the unembed, post-layerNorm (can't be unfolded RIP), sinusoidal positional embeddings. T5 might have other stuff.
- [ ] Keep in mind you might need a new config class to configure the entire architecture. If so, build it. 
- [ ] If you can pass the test for the model doing inference correctly and show that you can use .run_activation_cache and all the other methods that work on hooked transformer (within reason), then we're likely ready to merge. 

Some things to note:
- You should refine the to do list and add detail as you go. 
- Writing tests should make your life easier and create feedback cycles. It will also make it easier if multiple people want to collaborate on this. 
- A [previous PR ](https://github.com/neelnanda-io/TransformerLens/pull/82) attempted to provide BERT support is here. Unfortunately with drift it was too complicated to Merge. To avoid this, I recommend either syncing regularly or getting someone else to finish the PR if you can't. 
- Possibly quite useful, [a link to the original ARENA materials where we did this](https://github.com/callummcdougall/arena-v1/blob/main/w2d2/solutions_build_bert.py) without the hooks during the bootcamp

### Additional context

- [BERT paper](https://arxiv.org/abs/1810.04805)
- [BERT on huggingface](https://huggingface.co/bert-base-uncased)
- [Related T5 Model issue](https://github.com/neelnanda-io/TransformerLens/issues/94)

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by somvy**
Hey! 

I needed  EncoderDecoder (T5) models support  for my research, but unfortunately neither TransformerLens, nor nnsight had support for them. So i coded this one)

This PR is mostly based on #276 and HuggingFace implementation of [T5](https://github.com/huggingface/transformers/blob/v4.41.0/src/transformers/models/t5/modeling_t5.py)

Hope that it is enough to get started with encoder-decoder arhitectures, and merge into the lib. I will continue work on this. 

Will be glad to receive any comments 
:)

Loading example 

```python
from transformer_lens.HookedEncoderDecoder import  HookedEncoderDecoder

path = "results/t5_scratch_adamw_eos_3lay_8head_512d_2048dff_3e-4_400ep/"

#you can load  local custom model or official T5 weights from HF
path = "t5-small" 

hooked_model = HookedEncoderDecoder.from_pretrained(model_name=path)
```

# Description
* added HookedEncoderDecoder class
* added T5LayerNorm, T5Attention, T5Block
* changed signature of AbstractAttention.forward 
from `pos` to `kv_pos` in keys and values associated fields. This was needed for CrossAttention support
*  add model_state_dict loading from pretrained model
* added support of loading models from **local folders** 
    See more in `transformer_lens/loading_from_pretrained.py:669,1276,1437`  Unsure about this one. I need this feature for my research purposes, don't know if this should go into release. (But i believe there's some people, who might find this useful too.

Fixes #258  
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by bryce13950** [comment]
Very good! I will be reviewing this relatively quickly, but before I do that, can you do me a big favor, and add a demo, similar to the BERT demo you referenced in the other PR, so that the usage of this can be shown in detail for the future?

**Comment by somvy** [comment]
Thank you! 
Feel free to point to any weak spots in the code, im totally okey with it and will fix them as soon as possible)


Added a simple demo of t5 on how to run it with cache, visualise attn patterns and topk token logits

**Comment by bryce13950** [comment]
Thanks for the great work! I will hopefully have time to to review the whole thing this week. I have a few bugs that I need to fix, and hopefully nothing comes up after that.

**Comment by bryce13950** [comment]
Alright, so now that 2.0 has been put up I have sometime to review this. There is quite a bit changing with standards for code submissions in this project, which mainly revolve around improvements to unit testing, and a couple changing standards. What you have done so far looks pretty good, but given the changing standards I am going to request a few things to make sure we are moving closer towards the new standards.
That being said, the first thing I would like to check is to make sure you replace any asserts outside of tests with an if block and the throwing of an error. This is one of those things that isn't an official requirement yet for the code, but it will be very soon, and anything new coming in should be in this new format.
The second thing that I am going to ask is that we improve the test coverage a bit. The test you added is good, but I want to get some unit tests in here to test different bits of functionality in isolation a bit more. The two new components should have a unit test setup for them with the same directory structure. (e.g. a test at `tests/unit/test_t5_block.py`. Then in that unit test we need to test the individual groups of logic in isolation. This means that it might be a good idea to separate something like the different operation groups found in your forward function into smaller functions, so that we can focus in on those operations to ensure that they are being applied correctly.
If you have any questions on what I mean with this, let me know. I am happy to help you with improving the test coverage. It's pretty late here, so I don't have time to do any specific examples of what I mean, but I can do that either tomorrow or Monday if you want to wait for me to demonstrate exactly what I mean.

**Comment by somvy** [comment]
Hi!
Thank you for the feedback!
Made a quick fix to ensure asserts are replaced with ifs and errors

understand the issue with coverage, i will try ti cut some time to fix it on this week

Could you be more specific in 

> separate something like the different operation groups found in your forward function into smaller functions

you mean in `HookedEncoderDecoder.forward()` or in `T5Block`?

**Comment by bryce13950** [comment]
Honestly, after really taking a full look at what you did, I think we are probably fine. What I was referring to when I made the comment about separating functionality in your forward function, I was referring to your forward function on the T5Block, but after really having time to dig into your code today, I think what you are doing is fine. 

We are looking to find more ways to test blocks of functionality in isolation for the purpose of creating better unit tests. When it comes to doing that task on some of the classes you added, we may break it down a little bit more, so we can run tests to specific parts of your forward without touching the rest of it, but for the purpose of getting this into a release, I don't think there will be anything else requested of you. 

I will mess with your demo to verify a few last minute things tomorrow, and hopefully I will get this closed out at the same time. Thank you very much for your contribution here. I will let you know if anything else comes up, but I wouldn't expect anything.
